[33기 천은별] Add: 메뉴 페이지 리뷰 반영

eunbyulcheon opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [x] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 메뉴 페이지 - 상품 리스트 페이지, 상품 디테일 페이지 통신, 검색 기능 

<br />

## :: 구현한 사항 설명 
-  query parameter를 이용해 상품 카테고리마다 서버에서 메뉴 리스트를 불러온다. 
- path parameter를 이용해 메뉴 리스트 페이지에서 상품 디테일 페이지로 이동한다. 
- query parameter를 이용해 상품 검색 기능을 백엔드와 통신한다. 
- 상품 디테일 페이지에 closeHandler 버튼 구현

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- 백엔드와 통신. query, path parameter 사용 방법. 

<br />

## :: 기타 질문 및 특이 사항

eunbyulcheon committed:
Add: Product Page Layout (under Menu Page) - 초기 레이아웃

yeonuk-hwang commented on src/pages/Menu/Menu/Menu.js:
```diff
@@ -0,0 +1,7 @@
+  return <div>temp</div>;
+};
+
+export default temp;
```
불필요한 컴포넌트 같은데 추가하신 이유가 있나요??

yeonuk-hwang commented on src/pages/Menu/Product/AddToCart.js:
```diff
@@ -0,0 +1,55 @@
+  faPlus,
+  faX,
+  faHeart,
+} from '@fortawesome/free-solid-svg-icons';
```
```suggestion
import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faCircleQuestion,
  faMinus,
  faPlus,
  faX,
  faHeart,
} from '@fortawesome/free-solid-svg-icons';
import './AddToCart.scss';
```

import 순서 맞춰주세요~!!

yeonuk-hwang commented on src/pages/Menu/Product/AddToCart.scss:
```diff
@@ -0,0 +1,127 @@
+
+$borderColor: #c4c4c4;
+
+// ----------
```
주석 삭제해주세요!

yeonuk-hwang commented on src/pages/Menu/Product/AddToCart.scss:
```diff
@@ -0,0 +1,127 @@
+      display: flex;
+      justify-content: space-between;
+      align-items: center;
+      font-size: 1.1em;
```
- 전체적으로 깔끔하게 CSS 잘 작성해주셨네요 👍 , 다만 몇가지 속성 순서가 정렬되지 않은 부분이 보입니다!
- CSS 속성은 레이아웃에 영향을 많이 주는 순서대로, 인접 속성끼리 묶어서 
작성해주는게 좋습니다. 
- 일전에 안내해드린 [css 속성순서 컨벤션](https://study.wecode.co.kr/session/content/193) 확인하시고 지켜서 작성해주세요

yeonuk-hwang commented on src/pages/Menu/Product/Product.js:
```diff
@@ -0,0 +1,106 @@
+  faLocationDot,
+} from '@fortawesome/free-solid-svg-icons';
+import './Product.scss';
+import AddToCart from './AddToCart';
```
여기도 import 순서 맞춰주세요!

yeonuk-hwang commented on src/pages/Menu/Product/Product.js:
```diff
@@ -0,0 +1,106 @@
+            className="smallImage"
+            src="/images/menu/sweetPumpkinCreamPasta5.jpg"
+            alt="cream pasta"
+          />
```
반복되는 UI로 보이네요, 반복되는 UI는 하드코딩 하는 것 보다 Array.map 메서드를 활용하면 코드가 깔끔해지고, 유지보수에 용이해집니다.

yeonuk-hwang commented on src/pages/Menu/Product/Product.js:
```diff
@@ -0,0 +1,106 @@
+              &nbsp;&nbsp;
+              <FontAwesomeIcon icon={faLocationDot} title="새창" />
+              &nbsp;
+              <p className="emphasize">배송가능여부 조회</p>
```
공백들 같은 경우는 &nbsp 기호를 이용하는 것 보다 css를 이용해서 조정해주세요!

yeonuk-hwang commented on src/pages/Menu/Product/Product.js:
```diff
@@ -0,0 +1,106 @@
+          <div className="deliveryPriceBox">
+            <h4 className="gray">배송비</h4>
+            <p>
+              3,000원{' '}
```
auto formatting 되는 과정에서 {' '} 문자가 잘못 들어간 것 같네요 삭제해주세요!!

push 하고 나서 항상 PR의 File Changed 탭 확인해서 내가 작성한 코드에 실수는 없는지 의도한 작업이 맞는지 확인하는 습관을 들이시는게 좋습니다!

yeonuk-hwang reviewed (changes requested):
"select"와 "option" element들의 디폰트 스타일을 바꾸려면 스타일 값을 다 비활성화하고 새로운 스타일을 다 적용해주는 방법밖에 없는 건가요?
=> 네, 안타깝게도 select과 option은 스타일을 커스텀하기 제한된 부분이 있어서, 일부 우회하는 방법(기본 select tag를 안보이게 처리하고, 그 위에 다른 요소를 덮어씌우는 등)들을 통해서 커스텀을 진행할 수 밖에 없습니다 😢 

리뷰 확인 및 반영해주세요~!

yeonuk-hwang commented:
추가기능 구현중이신 부분은 어떤 부분일까요?

yeonuk-hwang commented:
은별님 수정반영 하고 계신가요?? 다른 작업보다 리뷰 반영이 먼저 되어야지 master에 브랜치들이 계속 통합될 수 있습니다!!

eunbyulcheon committed:
Modify: Product Page Reviewed & Add: Menu Page Layout

yeonuk-hwang commented on src/pages/Menu/Menu/Menu.js:
```diff
@@ -0,0 +1,61 @@
+        <Order />
+      </div>
+      <div className="contentDisplay">
+        {menuDisplay.map(
+          ({
+            id,
+            theme,
+            name,
+            description,
+            servings,
+            cookTime,
+            prepTime,
+            spice,
+            price,
+            totalReviews,
+            image,
+          }) => (
+            <MenuDisplay
+              key={id}
+              theme={theme}
+              name={name}
+              description={description}
+              servings={servings}
+              cookTime={cookTime}
+              prepTime={prepTime}
+              spice={spice}
+              price={price}
+              totalReviews={totalReviews}
+              image={image}
+            />
+          )
+        )}
```
```suggestion
        {menuDisplay.map((menu) => (
            <MenuDisplay
              key={id}
              {...menu}
            />
          )
        )}
```

만약 한 객체 안에 있는 모든 property를 그대로 props로 넘기고 싶다면 {...object} 형태로 작성해주면, key={value} 형태로 모든 객츼의 property가 props로 전달되는 JSX 문법이 있습니다.

예시 )
```js
const object = {name:"yeonuk", gender:"male"};
const {name, gender} = object;

<People name={name} gender={gender} />
```
```js
const object = {name:"yeonuk", gender:"male"};

<People {...object} />
```

그래서 지금 작성하신 것과 동일한 결과를 간단하게 할 수 있습니다, 다만, 객체안의 모든 프로퍼티가 다 넘어가게 되므로 반드시 모든 프로퍼티를 넘기고 싶을때만 사용하는게 좋습니다!! 만약 모든 프로퍼티를 넘길 필요가 없는데도 위와 같이 작성하게 된다면 자식 컴포넌트에서 필요 이상으로 많은 props를 알게되므로 좋지 않습니다! 상황에 맞게 적절하게 사용해주세요!

eunbyulcheon replied:
영욱님 리뷰 감사합니다!! 코밋 자주하고 리뷰 반영하겠습니다!!

yeonuk-hwang replied:
이름을 바꿔버리시다니.. 😭

yeonuk-hwang replied:
이부분 리뷰 반영해주세요!

eunbyulcheon replied:
연욱님~ 백엔드에서 받은 데이터 중에 menuDisplay.map()에 적용한 키들은 전부가 아닌데 그래도 말씀해주신 형식으로 바꿔도 되나요?

eunbyulcheon replied:
> 

죄송해요ㅜㅜ 오타예요!!!

yeonuk-hwang commented on src/pages/Menu/Menu/components/Category.scss:
```diff
@@ -0,0 +1,62 @@
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+}
```
여러 파일에서 공통적으로 활용할 것 같은 mixin이네요 공통적으로 활용할 mixin들은 variables.scss에 위치시키는게 좋습니다!

yeonuk-hwang reviewed (changes requested):
고생하셨습니다 리뷰 확인 및 반영해주세요!
- 리뷰 코멘트에 대해서 수정 하셨으면 이모지나 답글로 수정한 것 알 수 있게 리액션 해주세요!
- commit을 크게 두개로만 나눠서 하신 것 같은데 작업 중간중간에 자주 하시는 게 좋습니다! 커밋 안하고 파일 날라가면 못살리기에 중간중간에 자주 커밋 남겨주세요! 작업 중 불필요한 커밋(WIP(Woriking In Progress) commit이라고 통칭 많이 부릅니다, 그래서 커밋 메세지도 WIP 등으로 간략하게 남기기도 하고요)이 너무 많이 남는 것이 걱정될 수 있는데 2차 프로젝트 기간에는 여러 커밋들을 하나로 합치는 것도 학습하기 때문에 지금부터 커밋 자주하는 습관 들이시는게 좋습니다!
- 컨플릭도 해결해주세요!

yeonuk-hwang commented:
package-lock.json의 conflict는 아래 코멘트 참고해주세요
- lock 파일은 패키지 변화 히스토리 트리가 담겨있어서 프로젝트 버전이랑 함께 관리 되어야 합니다. lock 파일이 없다면 현재 프로젝트에서 사용하고 있는 버전이 아닌 최신 버전이 계속 설치됩니다.
- lock.json 파일 컨플릭트 해결 방법
    1. package.json 파일의 컨플릭트 해결
    2. `npm install --package-lock-only`
- 참고 링크 : [https://dltlabs.medium.com/package-json-vs-package-lock-json-c8d5deba12cb](https://dltlabs.medium.com/package-json-vs-package-lock-json-c8d5deba12cb)

eunbyulcheon committed:
Modify: Product, Menu Page Layout & Add: Category Function 초기

eunbyulcheon committed:
Merge with Main Page

eunbyulcheon committed:
Modify: Main Page - Product details

eunbyulcheon committed:
Add: galley.json Mock Data

eunbyulcheon committed:
백앤드와 통신: Products_list dispatch

eunbyulcheon committed:
Add: useState to category button on click

yeonuk-hwang commented on src/pages/Menu/Menu/components/category/Category.js:
```diff
@@ -0,0 +1,49 @@
+import './Category.scss';
+
+const Category = () => {
+  const [isCheckboxOpen, setIsCheckboxOpen] = useState(false);
+  const [isSearchboxOpen, setIsSearchboxOpen] = useState(false);
```
만약 둘 중 하나만 열려야 하는 상황이라면 state 하나로도 충분히 처리할 수 있어 보이고 그게 더 좋을 것 같네요!!
지금은 만약 버튼이 하나 더 추가된다면 state도 마찬가지로 하나 더 늘어나야 하는 상황으로 보입니다!

두개에서 늘어날 일이 없다면 하나의 boolean으로, 아니라면 특정한 string 등으로 표현하는것도 좋을 것 같네요

예시
```js
const [currentTab, setCurrentTab] = useState("theme");
```

eunbyulcheon replied:
```
<div className="flexRight">
          {isCheckboxOpen ? <Checkbox /> : null}
          {isSearchboxOpen ? <Searchbox /> : null}
 </div>
```


연욱님이 알려주신 대로 수정해봤었는데 윗 부분에 올려드린 것처럼 버튼마다 띄어주는 컴포넌트가 달라서 하나의 state로 묶으면 두 컴포넌트가 겹쳐 나오는 문제점이 있었습니다ㅠㅠ

yeonuk-hwang reviewed (commented):
리뷰 확인 및 반영해주세요!

yeonuk-hwang commented:
이 브랜치 어느정도 마무리 해도 될 것 같은데 추가기능구현은 어떤 항목인가요??

eunbyulcheon committed:
Add: Fetched data in Product.js

eunbyulcheon committed:
Modify: Combine Category Tab event into one

eunbyulcheon committed:
"Add: OnClick Event that triggers title change on each category display"

A State was created.
A function, titleHandler, was passed down all the way down to the lowest child component, Checkbox.
A variable, title, was passed down to another child component, Order.
When the buttons in Checkbox component receive a click, the title in the Order component changes, keeping it consistent with the category name.

eunbyulcheon committed:
"Add: functions in AddToCart.js"

- Added: closeHandler function - closes the "Option" div.
- Added: useState on the "Option" div - when clicked opens up.
- Added: increase and decrease quantity functions

eunbyulcheon committed:
Add: Connecting Menu Page with Product Detail page - Path param

yeonuk-hwang commented:
컨플릭 해결해주세요!

eunbyulcheon committed:
"Modify: Path Parameter, Navigate from Menu page to Product Detail Page"

Live Review

eunbyulcheon committed:
Conflict Resolved - variables.scss

eunbyulcheon committed:
Modify: Use of Path Parameter, useNagivate, and param.id with mock data

eunbyulcheon committed:
"Add: functions in AddToCart.js and ProductCarousel.js"

- onCLick event -> quantity affects total price of purchase.
- Carousel -> images slide (unfinished)

eunbyulcheon committed:
Use of navigate and location to request 'theme', 'sort' data from server

eunbyulcheon committed:
"Modify: Fetch data fom server's api"

eunbyulcheon commented on src/pages/Menu/Menu/Menu.js:
```diff
@@ -0,0 +1,93 @@
+      filterItem.themeValue ? `themeId=${filterItem.themeValue}` : ''
+    }${filterItem.sortValue ? `&sort=${filterItem.sortValue}` : ''}`;
+    navigate(queryString);
+  }, [filterItem]);
```
Dependency Array가 있어야 매 렌더링마다 콜백함수가 실행이 되서 꼭 필요합니다.

yeonuk-hwang replied:
이 부분은 의존성 배열에 navigate를 넣어주면 될 것 같네요!!
***
의존성 배열 동작이랑 어떻게 설정해야 되는지에 대해서 이해하기 위해서는 클로저, primitive type, reference type등 다양한 개념의 이해가 필요합니다.
useEffect에 관련되서 아래의 잘 정리된 아티클 하나 공유드립니다!
당장 이해안되더라도 나중에 찬찬히 읽어보고 찾아보시면 좋을것 같습니다!
[useEffect 완벽 가이드](https://www.rinae.dev/posts/a-complete-guide-to-useeffect-ko)
***
관련되서는 위 코멘트 참고해주세요! 저 내용 처음부터 모두 이해하기는 어려울거라서 아카이브 해뒀다가 필요할 때 계속 읽어보면서 차근차근 이해하는 방식으로 학습하시면 좋을 것 같습니다!

yeonuk-hwang commented:
은별님 PR 제목은 제일 최근 작업 내용이 아니라(라이브 리뷰 반영 등) 브랜치에서의 작업 목적 한줄 요약으로 작성해주세요!!
`Add: 메뉴 페이지` 이런 식으로요!

eunbyulcheon committed:
"Add: Search 기능 구현 백엔드와 통신 완료"

eunbyulcheon committed:
"Modify: Product Detail page"

- Show/unshow button on Options.

eunbyulcheon committed:
Modify: Conflict
